imsim/photon_ops.py

```diff
@@ -135,12 +119,248 @@ def applyTo(self, photon_array, local_wcs=None, rng=None):
+        `imsim.atmPSF.AtmosphericPSF` which stores sampled pupil
+        locations in photon_array.pupil_u and .pupil_v.
+
+        Parameters
+        ----------
+        photon_array:   A `PhotonArray` to apply the operator to.
+        local_wcs:      A `LocalWCS` instance defining the local WCS for the current photon
+                        bundle in case the operator needs this information.  [default: None]
+        rng:            A random number generator to use if needed. [default: None]
+        """
+
+        xy_to_v = XyToV(local_wcs, self.icrf_to_field, self.sky_pos)
+        # Convert xy coordinates to a cartesian 3d velocity vector of the photons
+        v = xy_to_v(photon_array.x, photon_array.y)
+
+        # Adjust for refractive index of air
+        wavelength = photon_array.wavelength * 1e-9
+        n = self.telescope.inMedium.getN(wavelength)
+        v /= n[:, None]
+
+        x, y = photon_array.pupil_u, photon_array.pupil_v
+        v = apply_diffraction_delta(
+            np.c_[x, y],
+            v,
+            photon_array.time,
+            wavelength,
+            lat=self.latitude,
+            az=self.azimuth,
+            alt=self.altitude,
+            geometry=LSST_SPIDER_GEOMETRY,
+            distribution=self.diffraction_rng,
+        )
+        photon_array.x, photon_array.y = xy_to_v.inverse(v)
+
+    def __str__(self):
+        return f"imsim.{type(self).__name__}()"
 
     def __repr__(self):
         return str(self)
 
 
+def specified_opt_kwargs(kwargs: dict, keys: list) -> dict:
+    """Extracts keys from kwargs, whose values are not None."""
+    return {key: kwargs.get(key) for key in keys if kwargs.get(key) is not None}
+
+
+def photon_op_type(identifier: str):
+    """Decorator which calls RegisterPhotonOpType on a PhotonOp factory,
+    defined by a function deserializing a PhotonOp from a dict.
+
+    The decorated function should have the following signature
+    > def deserializer(config, base, logger): ...
+
+    config, base and logger will receive the arguments of a call to
+    galsim.config.PhotonOpBuilder.buildPhotonOp.
+    """
+
+    def decorator(deserializer):
+        class Factory(PhotonOpBuilder):
+            """Build a PhotonOp generated by a deserializer.
+
+            Returns:
+                 the constructed PhotonOp object.
+            """
+
+            def buildPhotonOp(self, config, base, logger):
+                return deserializer(config, base, logger)
+
+        RegisterPhotonOpType(identifier, Factory())
+        return deserializer
+
+    return decorator
+
+
+def config_kwargs(config, base, cls):
+    """Given config and base, extract parameters (including optionals)
+    and optional parameters.
+    """
+    req, opt, single, _takes_rng = get_cls_params(cls)
+    kwargs, _safe = GetAllParams(config, base, req, opt, single)
+    opt_kwargs = specified_opt_kwargs(kwargs, cls._opt_params.keys())
```

Comment:
I think there must not be any test that runs this function, since I'm pretty sure there is no chance this could work.  `GetAllParams` cannot handle something with type `batoid.Optic`.  That should not be in the `_req` dict.  It needs to be created separately.  Presumably with the `load_telescope` function you made in batoid_utils.

Reply:
I also don't understand the point of the `specified_opt_kwargs` function.  Seems like what you are doing with that could just be done with the regular `kwargs` rather than doing this bit to separate out the optional kwargs.  Anyway, I suspect you mostly just need a test that runs this function so you can get this working correctly.

Reply:
After the split of the photon ops, we actually dont need it anymore. But before, we had `use_diffraction` with a default value of `True`. But when we did not specify this value in the config, the parameter was filled with a `None` which evaluated to `False` overriding the default of `True`.
Therefore I introduced this, to protect myself from that behavior (I just removed all values with `None` from `kwargs` assuming they were not specified in the config).
This is btw also something, I would put on my wish list for GalSim. Like this, it is a bit of a pitfall.

Reply:
I don't think that's right. If it's not in the config, it won't be in kwargs at all. There's nothing in GalSim, I don't think, that inserts a default value of None.  We have other places with default value that aren't None which work fine.

Reply:
But there is something which injects a `None` at this place.
You can verify yourself:
Go back to a84a7fa3031dc5ad204858e8e03619f4a202de5a (thats one commit before ad2bd07008bb7dfa14e52680eb2cdfd868d84ab7, which introduced my "hack"). There we also still have `use_diffraction`.

If I then run imsim with `galsim 1-star-near-zenith.yaml`, I end up with `use_diffraction=None`.

`1-star-near-zenith.yaml`:

```yml
---
modules: [imsim]
template: imsim-config

input.instance_catalog.file_name: 1-star-near-zenith.txt
input.opsim_meta_dict.file_name: 1-star-near-zenith.txt
input.tree_rings.only_dets: [R22_S11]

image.sky_level: 0.000001

gal:  # Override object in cat with this
  type: DeltaFunction
  sed:
    file_name: vega.txt
    wave_type: nm
    flux_type: flambda
    norm_flux_density: 1e5
    norm_wavelength: 500

output.dir: output/1-star-near-zenith
output.readout: ""
output.det_num:
  type: List
  items: [94]
output.nfiles: 1
output.truth: ""
```

`1-star-near-zenith.txt`:

```
#comment
rightascension 63.3
declination -30.14463
mjd 59797.2854090
altitude 89.9
azimuth 73.7707957
filter 2
rotskypos 69.0922930
rottelpos 1.000
dist2moon 145.1095257
moonalt -11.1383568
moondec -18.7702120
moonphase 59.6288830
moonra 230.9832941
nsnap 2
obshistid 161899
seed 161899
seeing 0.7613760
sunalt -59.1098785
vistime 33.0000000
object MS_567_8a 63.35000000000000 -30.1 14.0 starSED/phoSimMLT/lte033-4.5-1.0a+0.4.BT-Settl.spec.gz 0 0 0 0 0 0 point none CCM 0.0635117705 3.1
```

Reply:
I tried to verify by adding a print statement after `GetAllParams`, and kwargs is the following:
```
kwargs =  {'band': 'r', 'boresight': coord.CelestialCoord(coord.Angle(1.1047934165124105, coord.radians), coord.Angle(-0.5261230452954583, coord.radians)), 'telescope': <batoid.optic.Optic object at 0x177e849a0>}
```
There is no `use_diffraction: None` in there.  Did you intend me to do something different to verify?  Maybe I'm not understanding the problem you had.

For reference:
```
(lsst-scipipe-4.0.1) Mike@Saxophone:~/LSSTDESC/imSim[(HEAD detached at a84a7fa)*] $ git status
HEAD detached at a84a7fa
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   imsim/photon_ops.py

no changes added to commit (use "git add" and/or "git commit -a")
(lsst-scipipe-4.0.1) Mike@Saxophone:~/LSSTDESC/imSim[(HEAD detached at a84a7fa)*] $ git diff
diff --git a/imsim/photon_ops.py b/imsim/photon_ops.py
index a5fdb07..d253ab2 100644
--- a/imsim/photon_ops.py
+++ b/imsim/photon_ops.py
@@ -165,7 +165,10 @@ class LsstOpticsFactory(PhotonOpBuilder):
 
     def buildPhotonOp(self, config, base, _logger):
         req, opt, single, _takes_rng = get_cls_params(LsstOptics)
         kwargs, _safe = GetAllParams(config, base, req, opt, single)
+        print('kwargs = ',kwargs)
 
         opt_kwargs = {key: kwargs.get(key) for key in LsstOptics._opt_params.keys()}
         shift_optics = opt_kwargs.pop("shift_optics")
```

Reply:
I have to apologize! I only looked in the constructor of `LsstOptics` and found `use_diffraction` there.
I did not even have a look at the line (visible in your diff)
```py
opt_kwargs = {key: kwargs.get(key) for key in LsstOptics._opt_params.keys()}
```
Which actually adds the `None`.
So big sorry about the wrong accusation, this was introduced by me.

Reply:
OK. So the `specified_opt_kwargs` is gone now.
Currently, I also am not retrieving `telescope` from `kwargs`, but from `base`.
`telescope=base["_telescope"]`. Like this things seem to work (at least when running a simulation from `galsim`).

But I agree, that adding `"telescope"` to `_req_params` is probably not the clean way.
Is using `telescope=base["_telescope"]` a proper way? Or should the telescope be really built via `load_telescope`?

Regarding tests: Do you have functionality to populate `base` and `config` (e.g. according to a yaml config file represented by a `dict`)? Or has that to be done manually?

Reply:
> Is using telescope=base["_telescope"] a proper way? Or should the telescope be really built via load_telescope?

I believe `base['_telescope']` is made by the BatoidWCS.  So if we use that, then we are implicitly requiring that we always use that WCS whenever we want to use these photon ops.  Maybe that's ok.  But probably better practice would be to check for the cached `_telescope`, and if it's not present, then make it (and cache it).

Reply:
> Do you have functionality to populate base and config (e.g. according to a yaml config file represented by a dict)? Or has that to be done manually?

There are tests in some other imsim test files that test config dicts.  See test_flats, test_psf, and test_batoid_wcs for instance.  But you can also use `galsim.config.ReadConfig` to read in a config dict from a file if that's easier.  The GalSim tests of the config processing use both approaches.

Reply:
ok. I added some tests and refactored `load_telescope`